File: Two Sum - Difference equals to target/Javascript/solution_4.js
// https://www.lintcode.com/problem/610/

function twoSum7(nums, target) {
  let len = nums.length;

  let map = new Map();
  for (let i = 0; i < len; i++) {
    map.set(nums[i], i);
  }

  for (let i = 0; i < len; i++) {
    let num1 = nums[i];
    let num2 = num1 + Math.abs(target);
    if (map.has(num2) && map.get(num2) !== i) {
      return [num1, num2];
    }
  }

  return [];
}

let nums = [0, 5, 7];
let target = -2;
console.log(twoSum7(nums, target));

File: Two Sum - Difference equals to target/Rust/src/solution_4.rs
// https://www.lintcode.com/problem/610/

use std::collections::HashMap;

pub fn two_sum7(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let nums_len = nums.len();
    let mut result = vec![];

    let mut hash = HashMap::with_capacity(nums_len);
    for i in 0..nums_len {
        let num1 = nums[i];
        hash.insert(num1, i);
    }

    for i in 0..nums_len {
        let num1 = nums[i];
        let num2 = num1 + target.abs();
        if hash.contains_key(&num2) && *hash.get(&num2).unwrap() != i {
            return vec![num1, num2];
        }
    }

    result
}